### ISSUE TITLE
Merge with NuBuild

### ISSUE THREAD

**Issue by NN---** [nuproj/nuproj#154]
There is a NuBuild project with similar concepts. ( https://github.com/bspell1/NuBuild )
I wish both projects are merged into one maintained project for NuGet packaging.

**Comment by AArnott** [comment]
I think consolidating would be a good thing as well. I've never used NuBuild so I can't comment for sure, but from reading their README I don't see any particularly unique and mandatory features that NuProj doesn't offer. Given NuProj is based on CPS and NuBuild *appears* to be based on MPFproj, I think NuProj may be the better platform on which to build the unified solution. 
Have you spoken with the NuBuild folks?

**Comment by NN---** [comment]
I find NuProj an easier therefore a better choice for me as a beginner package writer.
So other teammates would like something easier to use.

I opened an issue there https://github.com/bspell1/NuBuild/issues/63 .
I hope that project is not dead and we get an answer :)

**Comment by lmagyar** [comment]
I'm the guy who forked NuBuild v1.x to v2.x (https://github.com/lmagyar/NuBuild) and added nearly the same main features as in NuProj. :)

I started to play with NuProj (ie. **not** a comprehensive test), read some doc/QA, but minimal source reading. I've found these issues (maybe bug, maybe I misunderstood something, maybe new features), so a **quick** analysis/comparison:

1. NuProj doesn't calculate the minimal set of the dependencies (installed packages != dependencies), it's tricky, but NuGet command line contains a "minimal set walker", I've based on that algorithm and created a DependencyManager "god class" that finally spits out FrameworkReferences and DependencySets. It can handle multiple framework targets also.

2. NuProj doesn't include the correct frameworkAssemblies (eg. System.Messaging wasn't included during my quick test, but it is not part of a default C# project, so I think it should be included). 

3. When targeting multiple framework versions, only the lowest version is included in frameworkAssemblies, is it a feature, that NuProj checks whether the reference list is identical for different framework versions and adds only the lowest? Will this work, when you add the package to the higher version framework (I didn't test it)? It is possible, that different sets are required for different framework versions, so is this implementation correct?

4. Is it possible to use the referenced binaries' properties (id, version, author, etc)? It is important to maintain these information at only one place.

5. If it is possible to use the referenced binaries' properties, and there are multiple referenced binaries, but with different id/name, title, description, can you select which binary will be the source for the properties? This should be a property/attribute of the ProjectReference.

6. Can you limit the major version of the dependencies?

7. Can you copy files to the output folder (next to the .nupkg file)?

8. NuBuild v1.x legacy, I've never used it: you can name the .nupkg file without the version.

9. If the referenced project's output is an .exe, NuProj should include it, the target folder in the package (eg. /lib or /tools) should be an optional property/attribute of the ProjectReference or a global setting.

10. To specify the output dir (and make it depending on Release/Debug configuration), would be useful.

11. To specify build events, would be useful. There are strange packaging situations, where we need them.

12. There is the .pp extension in NuGet for source code transformation, a'la pre-processing. We created the .ppp extension for pre-package-processing: before these text files are added to the package builder, $something$ variables can be replaced in them with the properties of the package (eg. $version$).

13. We add the NuBuild tool version to the project file. So if one developer edits the project with a newer version and adds some new properties or uses a new feature, but another developer (or the build server) uses an older version that eg. can't handle the new property, it causes a build error ie. forces an update. Version is updated when the project file / properties are edited.

14. Handling mixed AnyCPU, x86, x64 projects during MSBuild build was painful. MSBuild (even 12.0) can't determine the project build order from ProjectReference items (we have to use MSBuild task in this case on the referenced projects), but VS builds correctly. So we had to run MSBuild directly for the referenced projects before packaging, but only when building outside the VS. I see BuildingInsideVisualStudio conditions in your .targets file (simply searched for it, didn't test it), but I had no time to analyze your .targets. :) This feature is important, if you build & pack mixed eg. C++, hardware specific, crypto binaries.

15. Am I understand correctly: to use NuProj on a build server, eg. TFS you have to edit the .nuproj file manually? And how to install the NuProj.nupkg in a NuProj project? NuGet filters projects based on the project file extension, it is impossible to install it in to the NuProj project. It seems to me as magic, if it is not obvious, developers will screw it. I couldn't find clear description in the doc.

16. Strange: after installing the .vsix from an .msi, the Uninstall button should be grayed out in VS, you shouldn't be allowed to uninstall the .vsix without the uninstallation of the .msi. I've found InstalledByMsi="true" in the .vsixmanifest, strange.

17. NuProj install uses VS2013Setup/VS14Setup actions and those use "devenv.exe /setup". This is the reason, that the installation freezes at 99%. ;) The /updateconfiguration option is much faster, it's just a blink (it only touches the extensions.configurationchanged file in the Extensions folder). You need a CustomAction for this.

Why I forked NuBuild? The breaking change was that NuBuild v2.x uses C# project flavor and not standalone project system. This way you can install NuGet packages into these packaging projects (NuGet accepts the .csproj extension but filters out the .nuproj). It is more flexible than creating and sharing project templates. And project flavors have very nice Properties pages. :) But I understand, this is not an option at all.

How to continue? Currently my productivity is very limited due to my medical condition. So if you accept these issues as enhancements, I prefer to contribute only what is complex but requires lesser knowledge of NuProj internal logic (eg. better FrameworkReferences and DependencySets).

**Comment by kovalikp** [comment]
Just couple of quick comments:
1. :open_mouth: This looks interesting.
2. Currently frameworkAssemblies must be added manually and we don't have CPS support for them.
3. This appears to be a :bug:.
4. #68 needs to reviewed, updated, merged and extended functionality to other properties. 
5. See 4.
6. Yes. We have support for both .csproj to nuget and nuproj dependencies.
7. Nope, we don't have support for that.
8. What would be a use case for that feature?
9. You can specify, if the output of referenced project should go to lib, tools, ... folders.
10. Eeeyup. I think CPS can support that.
11. I prefer to use `BeforeTargets` and `AfterTargets` attribute to extend build functionality. 
12. Hrrrm, interesting idea.
13. I think it would be better to use NuProj nuget package in this case to have latest project system within solution.
14. I haven't noticed issues you described in NuProj.
15. NuGet does not have support for CPS projects we would require. I believe they are working on it. @AArnott will know more.
16. :bug: 
17. Should be easy to implement, even easier if I could just C&P it from somewhere.

**Comment by lmagyar** [comment]
May I suggest a quick triage:

**create new issue/bug**: 16
**create new issue/improvement**: (1-2-3) together, 7, 10, 12, 17
I can start to investigate: (1-2-3), 17

icebox: 11, 13, 15

already an issue: (4-5)
false alarm: 6, 9, 14
has no use case: 8

**Comment by AArnott** [comment]
Thanks for the very impressive list, @imagyar and @kovalikp for jumping on it so fast!
My quick read is that your interests are compatible with NuProj so I think the bugs/features that are missing from NuProj would make good additions. And it sounds like you two are already working on whittling down that list to an actionable set. Great. :)

**Comment by lmagyar** [comment]
Another false alarm: 16 - the uninstall button **is** grayed out, sorry, I've mixed NuProj with the Package Manager.

**Comment by kovalikp** [comment]
Correction to item 6 in my comment: version constraints for .csproj references are implemented in #129, which waits for review and merge.

**Comment by AArnott** [comment]
Perhaps at some point we can edit the original issue comment with the latest draft of the edited list of work items we'd want to do so we can discuss and ratify the plan overall. Then break it out into individual filed issues.

**Comment by lmagyar** [comment]
OK, I'm currently working on it.

**Comment by lmagyar** [comment]
### Recommended *issues* based on my previous comment:

1. Don't freeze installation at 99%
  * NuProj install uses VS2013Setup/VS14Setup actions and those use "devenv.exe /setup". This is the reason, that the installation freezes at 99%. ;) The /updateconfiguration option is much faster, it's just a blink (it only touches the extensions.configurationchanged file in the Extensions folder). You need a CustomAction for this.

2. Automatically calculate dependency minimal set and framework assemblies from project references
  * NuProj doesn't calculate the minimal set of the dependencies (installed packages != dependencies), it's tricky, but NuGet command line contains a "minimal set walker", I've based on that algorithm and created a DependencyManager "god class" that finally spits out FrameworkReferences and DependencySets. It can handle multiple framework targets also.

3. Add boolean `<LimitDependencyMajorVersion>` to project properties / Property Pages
  * This is an alternative to `<VersionConstraint>` #129 in `<ProjectReference>`, `<LimitDependencyMajorVersion>` automatically sets the constraint on all dependencies (on .nuproj project references and on installed NuGet packages in referred binary projects), you do not need to enter constant values. Eg. when the installed NuGet package version is 1.2.3 in a referenced binary project, the automatically limited major version constraint will be: [1.2.3,2.0.0)

4. Implement "Copy to Output Directory" functionality on project items
  * The property editor already maintains this value in the .nuproj file, the .targets file should execute it.

5. Add "Output Dir" to the Property Pages, and add it as configuration/platform dependent property
  * This is already in the project file, but not on UI.

6. Implement .ppp extensio for pre-package-processing
  * There is the .pp extension in NuGet for source code transformation, a'la pre-processing. Implement .ppp extension for pre-package-processing: before these text files are added to the package builder, $something$ variables can be replaced in them with the properties of the package (eg. $version$).

### And I've tested NuProj much deeper, some *improvements* to add:

1. Extend project reference's properties window
  * Add `<PackageDirectory>` and `<VersionConstraint>` #129 properties to the project reference's properties window. Currently there is only Identity and Path, it should be obvious that there are these properties without reading the doc and especially without editing the project file. Without this, what is the difference between text-editing the project file vs. a .nuspec file?

2. Group properties in the Property Pages
  * Eg. "Metadata", "General", "Config specific"

3. Add description to properties in the Property Pages
  * Eg. what does "No Package Analysis" mean?

4. Extend Property Pages with already existing properties
  * Add "Development Dependency" and "Language" metadata to Property Pages.
  * Add "Project defaults (template)" to Property Pages and let user choose from file-system. #56?

5. The "Require Licence Acceptance" property processed improperly
  * Even if the false value of "Require Licence Acceptance" is added to the project file, it can't overwrite the template's true value. A true value in the project file can overwrite a false value in the template.
  * The $(RequireLicenceAcceptance) macro has constant False value in the macro editor and replaces as an empty string during build, even when the template and the project file has true value.

6. Add "&lt;Edit...&gt;" option to the boolean "Require Licence Acceptance" property
  * What if somebody wants to use macros?

7. Treat empty properties in Property Pages as missing/null values
  * Does "&lt;inherit from parent or project defaults&gt;" option simply means, that the property will be deleted from the project file? And this option is only there, when the property is already in the project file, when you enter a value into an empty property, this option will be present only after Apply is pressed. But after you delete a property from the project file and reopen the property page, the field is simply empty. This is illogical and disturbing. I recommend 2 options:
    1. display missing properties always as "&lt;inherit...&gt;" option (ie. null = "&lt;inherit...&gt;")
    2. when a property is empty in the property pages, automatically delete it from the project file and forget this "&lt;inherit...&gt;" option, ie. empty=null=delete (I prefer this, because this is the natural and simple way to specify sg. in a .nuspec file also)
  * In "&lt;inherit from parent or project defaults&gt;", what does "parent" mean at all?

### And finally *questions*:

1. Why these metadata properties have to be specified in the project file and can't be inherited from the template? *id, version, title, authors, owners, description* #57?

2. Why these metadata properties can't be inherited from the template? *minClientVersion*

3. Why these metadata properties are not in the macro name/value pair list? *icon/project/licence url, summary, tag, release notes*

4. Why these metadata properties have no any metadata property values in the macro name/value list? They can't refer any other metadata property. *id, version, title, author, owner*
